### PR TITLE
Add tests and comments for lazy ESState implementation

### PR DESCRIPTION
**Pull request by sebgl**
Add some comments on all functions, and unit test ESState behaviour, by
relying on a mocked ES client.

I'm pretty sure this ESState thing will be improved and maybe handled
differently in the future, but in the meantime it's best to have it unit
tested correctly.

I hesitated to switch to a proper Mock tool for the ES Client, but I'm
actually quite happy with the current "manual mock" approach.

Relates #1287.

### CODE REVIEW

File: operators/pkg/controller/elasticsearch/driver/esstate.go [hunk 1]
```diff
@@ -71,12 +82,14 @@ func (n *lazyNodes) NodesInCluster(nodeNames []string) (bool, error) {
 
 // -- Shards allocation enabled
 
+// lazyNodes provides shards allocation information.
```
Comment (david-kow) [inline operators/pkg/controller/elasticsearch/driver/esstate.go]: godoc not matching

File: operators/pkg/controller/elasticsearch/driver/esstate.go [hunk 2]
```diff
@@ -13,19 +13,26 @@ import (
...
 	NodesInCluster(nodeNames []string) (bool, error)
+	// ShardAllocationsEnabled returns true if shards allocation are enabled in the cluster.
 	ShardAllocationsEnabled() (bool, error)
+	// GreenHealth returns true if the cluster health is currently green.
 	GreenHealth() (bool, error)
 }
 
+// LazyESState requests Elasticsearch for the requested information only once, at first call.
+// It is "lazy" in the sense it only calls Elasticsearch if required, and does not pre-populate the state.
 type LazyESState struct {
```
Comment (david-kow) [inline operators/pkg/controller/elasticsearch/driver/esstate.go]: It feels to me that lazy doesn't tell the full story (the part about cashing the result), but I don't have anything great to use instead.

Comment (sebgl) [reply operators/pkg/controller/elasticsearch/driver/esstate.go]: Yes, I agree. At least it's specified in the comment though.

Comment (pebrc) [reply operators/pkg/controller/elasticsearch/driver/esstate.go]: `MemoizingESState` memoization implies laziness and caching, not sure if that is a very intuitive name though...

Comment (sebgl) [reply operators/pkg/controller/elasticsearch/driver/esstate.go]: I had to Google it, it seems to match perfectly! Not very intuitive but at least you learn new vocabulary when reading the code? 😄 
I'm 👍 switching to that.

**Pull request merged**